### ISSUE TITLE
"eth_signTypedData" presents fields that do not appear in 'types' filed

### ISSUE THREAD

**Issue by forshtat** [MetaMask/metamask-extension#11457]
**Describe the bug**
When executing a call to 'eth_signTypedData', the expectation is that the message that is presented to the user and the data that will be signed must match exactly. In fact, what seems to happen is that MetaMask presents the entire 'message' field, including fields that do not appear in the 'types' field. If that is in fact what is happening, it seems to present some security risk as clients can be expecting to sign on some fields, such as "expirationBlock" and such, but in fact these fields are not part of EIP-712 signature created.

**Steps to reproduce (REQUIRED)**
Note that the pseudocode is a simplified version of affected production code:

1. Define an EIP-712 type structure:
```
dataToSign.type = [
  { name: 'operationA', type: 'uint256' },
  { name: 'operationB', type: 'uint256' }
]
``` 

2. Create an object:
```
dataToSign.message = {
  operationA: 123,
  operationA: 456,
  expirationBlock: 12777777
}
```
3. Execute

```
web3.currentProvider.send({
  method: 'eth_signTypedData_v4',
  params: [senderAddress, dataToSign],
  from: senderAddress
})
```

**Expected behavior**
Expectation: `expirationBlock` field does not appear in a signature popup. UPD: preferably, exception is raised indicating that structure does not match the supposed data type.
Actual: `expirationBlock` field appears as usual, but is not part of a signed data.

**Comment by Gudahtt** [comment]
I can confirm that any extra data missing from the types gets ignored in the final signed message. We sanitize the message object here by creating a new object, and only copying over defined properties: https://github.com/MetaMask/eth-sig-util/blob/main/src/sign-typed-data.ts#L360

>preferably, exception is raised indicating that structure does not match the supposed data type.

This would be nice! But it would be a breaking change. If we want to consider this, we'll want to give dapps plenty of time to fix their `eth_signTypedData` usage before we start throwing an error. We could start by doing some validation on the provider side and logging a warning at least though :thinking:.

I agree that rendering the field is confusing if it ends up getting ignored in practice. Maybe we could run `santiizeData` on the message here in the extension before we display it, ensuring that we only render exactly what gets signed.

**Comment by segun** [comment]
I can't seem to replicate this in Chrome.

At first I was getting an error that the method can not called without a callback, so I added a called link e this

```
web3.currentProvider.send({
  method: 'eth_signTypedData_v4',
  params: [senderAddress, dataToSign],
  from: senderAddress
}, (error, report) => { 	
    console.log("Error: ", error);
    console.log("Result: ", report); });
```

Then I got an error saying `params.data" must be a string.'`

**Comment by Gudahtt** [comment]
@segun here is an example of an `eth_signTypedData_v4` call: https://github.com/MetaMask/test-dapp/blob/1072816a6818bba745694c2fc966f95451bb1c18/src/index.js#L950

You can modify that message to include extra data to repro this issue.

**Comment by segun** [comment]
@Gudahtt the sanitizeData method only removes top level non EIP properties, what the issue described is a little different. It's talking about not showing message properties that are not defined in the types field.

Parsing the types field and making sure the message field matches exactly is I don't think a trivial thing to do and might actually be out of scope of what a wallet should do. The message type described above is simple, but you can have very complex message types like the one described in test-dapp.

```
      types: {
        EIP712Domain: [
          { name: 'name', type: 'string' },
          { name: 'version', type: 'string' },
          { name: 'chainId', type: 'uint256' },
          { name: 'verifyingContract', type: 'address' },
        ],
        Group: [
          { name: 'name', type: 'string' },
          { name: 'members', type: 'Person[]' },
        ],
        Mail: [
          { name: 'from', type: 'Person' },
          { name: 'to', type: 'Person[]' },
          { name: 'contents', type: 'string' },
        ],
        Person: [
          { name: 'name', type: 'string' },
          { name: 'wallets', type: 'address[]' },
        ],
      },
 ```

**Comment by Gudahtt** [comment]
Ah, good point. I thought it excluded extra top-level message properties, but it does not. And I agree that we'd want it to recursively exclude extra properties anyway.

`sanitizeData` could be updated to recursively filter out extra message properties. It would be a breaking change for that specific method, but it wouldn't break signing. These properties are already ignored during signing (there are unit tests that verify this).

It's not trivial, but it shouldn't be especially complicated. The message is guaranteed to be an entry in the `types` collection (the one listed as a `primaryType`), so that's where we can start. Then for each property, if that property's type is also in the `types` collection, sanitize that property with that type.
